File: sorting_algorithms/quick_sort/quicksort.c
#include <stdio.h>
#include <stdlib.h>

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int A[], int start, int end) {
    int pivot = A[end];
    int i = start - 1;
    for (int j = start; j < end; j++) {
        if (A[j] <= pivot) {
            i++;
            swap(&A[i], &A[j]);
        }
    }
    swap(&A[i + 1], &A[end]);
    return i + 1;
}

void quicksort(int A[], int start, int end) {
    if (start < end) {
        int pivot = partition(A, start, end);
        quicksort(A, start, pivot - 1);
        quicksort(A, pivot + 1, end);
    }
}

int main() {
    int A[] = {5, 4, 3, 2, 1};
    int n = sizeof(A) / sizeof(A[0]);
    quicksort(A, 0, n - 1);
    
    for (int i = 0; i < n; i++) {
        printf("%d ", A[i]);
    }
    printf("\n");

    return 0;
}

File: sorting_algorithms/quick_sort/quicksort.py
def partition(A, start, end):
    pivot = A[end]
    i = start - 1
    array = A[start:end + 1]
    for j in range(start, end):
        if A[j] <= pivot:
            i += 1
            A[i], A[j] = A[j], A[i]
    A[i + 1], A[end] = A[end], A[i + 1]
    return i + 1

def quicksort(A, start, end):
    if start < end:
        p = partition(A, start, end)

        quicksort(A, start, p - 1)
        quicksort(A, p + 1, end)

A = [5, 2, 4, 6, 1, 3]
quicksort(A, 0, len(A) - 1)
print(list(A))